src/libmongoc/tests/test-mongoc-bulk.c

```diff
@@ -5011,6 +5011,35 @@ test_bulk_write_multiple_errors (void *unused)
    mongoc_client_destroy (client);
 }
 
+// test_bulk_write_set_client_after_operation is a regression test for
+// CDRIVER-4665.
+static void
+test_bulk_write_set_client_after_operation (void)
+{
+   mongoc_client_t *client;
+   mongoc_bulk_operation_t *bulk;
+   bson_t reply;
+   bson_error_t error;
+   bool ok;
+
+   client = test_framework_new_default_client ();
```

Comment:
Does the regression test actually need a `client`, or is this just to demonstrate that you can ultimately execute the bulk operation.

The most minimal test case would be to create a bulk via `mongoc_bulk_operation_new()` (never calling `mongoc_bulk_operation_set_client()`) and then add one operation of each type w/o any assertion errors. I agree the additional operations are redundant given how the opts functions are generated.